core/domain/subtopic_page_domain.py

```diff
@@ -0,0 +1,204 @@
+# coding: utf-8
+#
+# Copyright 2018 The Oppia Authors. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS-IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Domain objects for the pages for subtopics, and related models."""
+
+import copy
+
+from constants import constants
+from core.domain import html_cleaner
+from core.platform import models
+import utils
+
+(topic_models,) = models.Registry.import_models([models.NAMES.topic])
+
+SUBTOPIC_PAGE_PROPERTY_HTML_DATA = 'html_data'
+
+CMD_ADD_SUBTOPIC = 'add_subtopic'
+CMD_CREATE_NEW = 'create_new'
+CMD_DELETE_SUBTOPIC = 'delete_subtopic'
+# These take additional 'property_name' and 'new_value' parameters and,
+# optionally, 'old_value'.
+CMD_UPDATE_SUBTOPIC_PAGE_PROPERTY = 'update_subtopic_page_property'
+
+
+class SubtopicPageChange(object):
+    """Domain object for changes made to subtopic_page object."""
+
+    SUBTOPIC_PAGE_PROPERTIES = (SUBTOPIC_PAGE_PROPERTY_HTML_DATA,)
+
+    OPTIONAL_CMD_ATTRIBUTE_NAMES = [
+        'property_name', 'new_value', 'old_value', 'name', 'subtopic_id',
+        'topic_id'
+    ]
+
+    def __init__(self, change_dict):
+        """Initialize a SubtopicPageChange object from a dict.
+
+        Args:
+            change_dict: dict. Represents a command. It should have a 'cmd'
+                key, and one or more other keys. The keys depend on what the
+                value for 'cmd' is. The possible values for 'cmd' are listed
+                below, together with the other keys in the dict:
+                - 'update_topic_property' (with property_name, new_value
+                and old_value)
+
+        Raises:
+            Exception: The given change dict is not valid.
+        """
+        if 'cmd' not in change_dict:
+            raise Exception('Invalid change_dict: %s' % change_dict)
+        self.cmd = change_dict['cmd']
+
+        if self.cmd == CMD_UPDATE_SUBTOPIC_PAGE_PROPERTY:
+            if (change_dict['property_name'] not in
+                    self.SUBTOPIC_PAGE_PROPERTIES):
+                raise Exception('Invalid change_dict: %s' % change_dict)
+            self.property_name = change_dict['property_name']
+            self.new_value = copy.deepcopy(change_dict['new_value'])
+            self.old_value = copy.deepcopy(change_dict['old_value'])
+            self.id = change_dict['subtopic_id']
+        elif self.cmd == CMD_CREATE_NEW:
+            self.topic_id = change_dict['topic_id']
+        else:
```

Comment:
Just to check: when a subtopic is deleted, is the subtopic page kept around in the datastore, or also deleted?

Reply:
Yes, I've added tests for that also in topic_services_test. When a subtopic page already in datastore is deleted, it is actually removed, but this happens similar to explorations and collections where in they are only marked as deleted and not completely removed.